### ISSUE TITLE
Rancher 1.2.0 cannot pull images from private registry

### ISSUE THREAD

**Issue by stephan2012** [rancher/rancher#6947]
**Rancher Version:**
1.2.0

**Docker Version:**
1.12.3

**OS and where are the hosts located? (cloud, bare metal, etc):**
RHEL 7.3, on-premise

**Setup Details: (single node rancher vs. HA rancher, internal DB vs. external DB)**
Single node Rancher, 3 nodes

**Environment Type: (Cattle/Kubernetes/Swarm/Mesos)**
Cattle

**Steps to Reproduce:**
Configure private regsitry (Nexus Repository Manager 3.1), try to pull image

**Results:** 
```
2016-12-06 10:10:20,666 ERROR [7b7a2f67-1fcf-4fb6-a3bc-3d3c84fbe44a:1486] [instance:92->instanceHostMap:92] [instance.start->(InstanceStart)->instancehostmap.activate] [] [ecutorService-5] [c.p.e.p.i.DefaultProcessInstanceImpl] Agent error for [compute.instance.activate.reply;agent=1]: Image [internal.registry.company.com:18443/nginx] failed to pull: Pulling repository internal.registry.company.com:18443/nginx 
2016-12-06 10:10:20,667 ERROR [7b7a2f67-1fcf-4fb6-a3bc-3d3c84fbe44a:1486] [instance:92] [instance.start->(InstanceStart)] [] [ecutorService-5] [i.c.p.process.instance.InstanceStart] Failed [1/2] to Starting for instance [92] 
2016-12-06 10:10:20,838 ERROR [7b7a2f67-1fcf-4fb6-a3bc-3d3c84fbe44a:1486] [instance:92->instanceHostMap:92] [instance.start->(InstanceStart)->instancehostmap.activate] [] [ecutorService-5] [c.p.e.p.i.DefaultProcessInstanceImpl] Agent error for [compute.instance.activate.reply;agent=1]: Image [internal.registry.company.com:18443/nginx] failed to pull: Pulling repository internal.registry.company.com:18443/nginx 
2016-12-06 10:10:20,838 ERROR [7b7a2f67-1fcf-4fb6-a3bc-3d3c84fbe44a:1486] [instance:92] [instance.start->(InstanceStart)] [] [ecutorService-5] [i.c.p.process.instance.InstanceStart] Failed [2/2] to Starting for instance [92] 
2016-12-06 10:10:21,418 ERROR [e4680aca-d67c-48f8-98ba-4743ab0aa3a6:1378] [service:17] [service.activate] [] [cutorService-11] [c.p.e.p.i.DefaultProcessInstanceImpl] Expected state running but got error: Image [internal.registry.company.com:18443/nginx] failed to pull: Pulling repository internal.registry.company.com:18443/nginx 
2016-12-06 10:10:21,430 ERROR [:] [] [] [] [cutorService-11] [.e.s.i.ProcessInstanceDispatcherImpl] Expected state running but got error: Image [internal.registry.company.com:18443/nginx] failed to pull: Pulling repository internal.registry.company.com:18443/nginx 
```

Note: Pulling images manually on the nodes works without any issues (with docker login first, of course).

Nexus records:

```
10.197.80.35 - - [06/Dez/2016:11:14:19 +0100] "GET /repository/docker-group/v2/ HTTP/1.1" 401 113 5
10.197.80.35 - - [06/Dez/2016:11:14:19 +0100] "GET /repository/docker-group/v1/_ping HTTP/1.1" 200 0 3
10.197.80.35 - - [06/Dez/2016:11:14:19 +0100] "GET /repository/docker-group/v1/repositories/nginx/images HTTP/1.1" 401 0 2
```

Looks like Rancher does not pass the configured registry credentials.

**Expected:**
Rancher pulls images from private regsitry as it did in 1.1.x.

**Comment by T-Grave** [comment]
I've managed to get this working by creating the registry credentials before adding any hosts to my environment.

**Comment by StrongMonkey** [comment]
Can you re-add your registry credential and try again?

**Comment by ekristen** [comment]
I'm seeing this behavior too.

I just installed Rancher 1.2.1 and I can't pull any private images from my private registry. The registry works just fine with the docker cli and everything else, but rancher isn't behaving properly.

It is attempting to talk to /v2/ and gets a 401, it appears to go fetch a token from the token login service but never seems to pass that information back, because the registry continues to through 401 errors, then it attempts to fallback to /v1 ...

**Comment by ekristen** [comment]
@StrongMonkey I've tried, it doesn't seem to work.

**Comment by StrongMonkey** [comment]
I did a PR in 1.2.1 https://github.com/rancher/agent/pull/112 and it should fix the problem. @stephan2012 Can you try 1.2.1 to see if that still happens?

**Comment by StrongMonkey** [comment]
@ekristen what private registry are you using? this issue should be addressed by 1.2.1.

**Comment by ekristen** [comment]
I'm using registry:2.5.1 from docker + an unreleased authentication service (but it implements the token authentication protocol outlined by the docker registry spec)

docker-compose, docker, all have zero problems interacting with this registry. 

The authentication service is written by me, I'm not 100% ready to release it although I plan to as open source, regardless, it implements this -- https://github.com/docker/distribution/blob/master/docs/spec/auth/token.md

Based on the logs it never authenticates and then it ultimately falls back to trying /v1/ endpoint stuff and gets a 404 and errors.

@StrongMonkey how can I better debug this?

**Comment by ekristen** [comment]
FWIW It "appears" that it attempts to authenticate, because I can see the request hit the auth server, and it gets a `token` back, but the registry logs show that all requests are unauthorized.

**Comment by ekristen** [comment]
@StrongMonkey looks like it was an issue on my side ... 

TL;DR - my proxy registered multiple bad services due to the https://github.com/rancher/rancher/issues/7016 happening and docker events not firing properly, causing some requests to get routed to nothing and causing a 404, thus making it appear that auth wasn't working. 

Explanation: How I got here was I originally installed the rancher-agent with swarmkit, after it attempted to put stuff into swarm mode and found it couldn't because live-restore was enable, this caused after a while the servers to become unstable (why I am not sure, but it seems like a lot of file handlers were being held open). Stopping docker and restarting didn't help, actually after the restart, I couldn't query docker anymore.  I had to reboot my servers which caused the *TL;DR* part of this mess. Furthermore, I still think that https://github.com/rancher/rancher/issues/7016 is a big issue, it causes containers to not be killed and exit properly and thus prevent certain docker events from firing.

**Comment by deniseschannon** [comment]
@stephan2012 Please upgrade to 1.2.2 and let me know if you are still facing issues. As @StrongMonkey has mentioned, there is an update in that version that should fix your issue.

**Comment by stephan2012** [comment]
@deniseschannon: With 1.2.2, Rancher pulls images from our private regsitry (Nexus 3) as expected. Thanks for fixing it.
